src/Datadog.Trace.ClrProfiler.Managed/Integrations/AWS/AmazonSQSIntegration.cs

```diff
@@ -0,0 +1,254 @@
+
+        private static async Task<T> InvokeAsyncInternal<T>(
+            object runtimePipeline,
+            object executionContext,
+            int opCode,
+            int mdToken,
+            long moduleVersionPtr)
+        {
+            Func<object, object, Task<T>> instrumentedMethod;
+            var runtimePipelineType = runtimePipeline.GetType();
+            var genericArgument = typeof(T);
+
+            try
+            {
+                instrumentedMethod =
+                    MethodBuilder<Func<object, object, Task<T>>>
+                    .Start(moduleVersionPtr, mdToken, opCode, InvokeAsyncMethod)
+                    .WithConcreteType(runtimePipelineType)
+                    .WithMethodGenerics(genericArgument)
+                    .WithParameters(executionContext)
+                    .WithNamespaceAndNameFilters(ClrNames.GenericTask, IExecutionContextTypeName)
+                    .Build();
+            }
+            catch (Exception ex)
+            {
+                Log.ErrorRetrievingMethod(
+                    exception: ex,
+                    moduleVersionPointer: moduleVersionPtr,
+                    mdToken: mdToken,
+                    opCode: opCode,
+                    instrumentedType: RuntimePipelineTypeName,
+                    methodName: InvokeAsyncMethod,
+                    instanceType: runtimePipelineType.AssemblyQualifiedName);
+                throw;
+            }
+
+            using (var scope = CreateScopeFromExecutionContext(executionContext))
+            {
+                try
+                {
+                    return await instrumentedMethod(runtimePipeline, executionContext).ConfigureAwait(false);
+                }
+                catch (Exception ex)
+                {
+                    scope?.Span.SetException(ex);
+                    throw;
+                }
+                finally
+                {
+                    // Because we need to wait for the return to decorate some values, we need to check here to ensure exceptions don't miss decorations
+                    if (scope?.Span != null)
+                    {
+                        AfterMethod(scope.Span, executionContext);
+                    }
+                }
+            }
+        }
+
+        private static Scope CreateScopeFromExecutionContext(object executionContext)
+        {
+            if (!Tracer.Instance.Settings.IsIntegrationEnabled(IntegrationName))
+            {
+                // integration disabled, don't create a scope, skip this trace
+                return null;
+            }
+
+            var tracer = Tracer.Instance;
+            Scope scope = null;
+
+            try
+            {
+                var serviceName = $"{tracer.DefaultServiceName}-{ServiceName}";
+                scope = Tracer.Instance.StartActive(OperationName, serviceName: serviceName);
+                var span = scope.Span;
+                span.Type = SpanTypes.Http; // TODO, is this right?
```

Comment:
Since Amazon SQS is a queuing system, it seems that we should have a new `SpanType.Queue`.

Reply:
I would agree if this code were acting as the queue, but it is only the client for the queue.

Reply:
This did make me notice that our operation name seems to differ from java's implementation.
https://github.com/DataDog/dd-trace-java/blob/00fe40f1fba5c0df65a8af80430feb2a3cba6eea/dd-java-agent/instrumentation/aws-java-sdk-2.2/src/test/groovy/AwsClientTest.groovy#L78

Reply:
In fact they seem to have a mapping of operation names: https://github.com/DataDog/dd-trace-java/blob/42b88c7b8ec99b8b9ac2a4d4dda90ebbbe9e5921/dd-java-agent/instrumentation/aws-java-sdk-1.11.0/src/main/java/datadog/trace/instrumentation/aws/v0/AwsSdkClientDecorator.java#L37

Reply:
This seems like a consistency issue then. I would check for RFCs that apply to naming operations if such an RFC exists. And maybe asking in the `apm-integrations` Slack channel.

Reply:
I worked with @tylerbenson to get a better idea of what Java is doing.

Reply:
Sounds good, thanks @colin-higgins !